Ingester crashes on exit

Spotted on one of the ingesters:
```
level=info ts=2018-12-14T09:13:39.379995933Z caller=gokit.go:36 msg="=== received SIGINT/SIGTERM ===\n*** exiting"
level=info ts=2018-12-14T09:13:39.382026879Z caller=loki.go:158 msg=stopping module=store
level=info ts=2018-12-14T09:13:39.382126813Z caller=loki.go:158 msg=stopping module=ingester
level=info ts=2018-12-14T09:13:39.382174645Z caller=lifecycler.go:443 msg="changing ingester state from" old_state=ACTIVE new_state=LEAVING
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0xbbfa8c]

goroutine 88517526 [running]:
github.com/grafana/loki/vendor/cloud.google.com/go/storage.(*Writer).open.func1(0xc077ed3860, 0xc15b1dc780, 0xc06cc148c0, 0xc231538280, 0x1, 0x1)
	/go/src/github.com/grafana/loki/vendor/cloud.google.com/go/storage/writer.go:118 +0xfc
created by github.com/grafana/loki/vendor/cloud.google.com/go/storage.(*Writer).open
	/go/src/github.com/grafana/loki/vendor/cloud.google.com/go/storage/writer.go:108 +0x471
```

--- a/pkg/loki/loki.go
+++ b/pkg/loki/loki.go
@@ -70,15 +70,12 @@ type Loki struct {
 	store       chunk.Store
 
 	httpAuthMiddleware middleware.Interface
-
-	inited map[moduleName]struct{}
 }
 
 // New makes a new Loki.
 func New(cfg Config) (*Loki, error) {
 	loki := &Loki{
-		cfg:    cfg,
-		inited: map[moduleName]struct{}{},
+		cfg: cfg,
 	}
 
 	loki.setupAuthMiddleware()
@@ -128,8 +125,6 @@ func (t *Loki) initModule(m moduleName) error {
 			return errors.Wrap(err, fmt.Sprintf("error initialising module: %s", m))
 		}
 	}
-
-	t.inited[m] = struct{}{}
 	return nil
 }
 
@@ -161,5 +156,4 @@ func (t *Loki) stopModule(m moduleName) {
 			level.Error(util.Logger).Log("msg", "error stopping", "module", m, "err", err)
 		}
 	}
-	delete(t.inited, m)
 }

--- a/pkg/loki/loki.go
+++ b/pkg/loki/loki.go
@@ -111,14 +111,19 @@ func (t *Loki) setupAuthMiddleware() {
 }
 
 func (t *Loki) init(m moduleName) error {
-	if _, ok := t.inited[m]; ok {
-		return nil
-	}
-
-	for _, dep := range modules[m].deps {
-		if err := t.init(dep); err != nil {
+	// initialize all of our dependencies first
+	for _, dep := range getDeps(m) {
+		if err := t.initModule(dep); err != nil {
 			return err
 		}
+	}
+	// lastly, initialize the requested module
+	return t.initModule(m)
+}
+
+func (t *Loki) initModule(m moduleName) error {
+	if _, ok := t.inited[m]; ok {
+		return nil
 	}
 
 	level.Info(util.Logger).Log("msg", "initialising", "module", m)
@@ -145,21 +150,23 @@ func (t *Loki) Stop() error {
 }
 
 func (t *Loki) stop(m moduleName) {
+	t.stopModule(m)
+	deps := getDeps(m)
+	// iterate over our deps in reverse order and call stopModule
+	for i := len(deps) - 1; i >= 0; i-- {
+		t.stopModule(deps[i])
+	}
+}
+
+func (t *Loki) stopModule(m moduleName) {
 	if _, ok := t.inited[m]; !ok {
 		return
 	}
-	delete(t.inited, m)
-
-	for _, dep := range modules[m].deps {
-		t.stop(dep)
-	}
-
-	if modules[m].stop == nil {
-		return
-	}
-
 	level.Info(util.Logger).Log("msg", "stopping", "module", m)
-	if err := modules[m].stop(t); err != nil {
-		level.Error(util.Logger).Log("msg", "error stopping", "module", m, "err", err)
+	if modules[m].stop != nil {
+		if err := modules[m].stop(t); err != nil {
+			level.Error(util.Logger).Log("msg", "error stopping", "module", m, "err", err)
+		}
 	}
+	delete(t.inited, m)
 }

--- a/pkg/loki/loki.go
+++ b/pkg/loki/loki.go
@@ -112,7 +112,7 @@ func (t *Loki) setupAuthMiddleware() {
 
 func (t *Loki) init(m moduleName) error {
 	// initialize all of our dependencies first
-	for _, dep := range getDeps(m) {
+	for _, dep := range orderedDeps(m) {
 		if err := t.initModule(dep); err != nil {
 			return err
 		}
@@ -122,10 +122,6 @@ func (t *Loki) init(m moduleName) error {
 }
 
 func (t *Loki) initModule(m moduleName) error {
-	if _, ok := t.inited[m]; ok {
-		return nil
-	}
-
 	level.Info(util.Logger).Log("msg", "initialising", "module", m)
 	if modules[m].init != nil {
 		if err := modules[m].init(t); err != nil {
@@ -151,7 +147,7 @@ func (t *Loki) Stop() error {
 
 func (t *Loki) stop(m moduleName) {
 	t.stopModule(m)
-	deps := getDeps(m)
+	deps := orderedDeps(m)
 	// iterate over our deps in reverse order and call stopModule
 	for i := len(deps) - 1; i >= 0; i-- {
 		t.stopModule(deps[i])
@@ -159,9 +155,6 @@ func (t *Loki) stop(m moduleName) {
 }
 
 func (t *Loki) stopModule(m moduleName) {
-	if _, ok := t.inited[m]; !ok {
-		return
-	}
 	level.Info(util.Logger).Log("msg", "stopping", "module", m)
 	if modules[m].stop != nil {
 		if err := modules[m].stop(t); err != nil {

--- a/pkg/loki/modules.go
+++ b/pkg/loki/modules.go
@@ -187,41 +187,38 @@ func listDeps(m moduleName) []moduleName {
 	return deps
 }
 
-// getDeps gets a list of all dependencies ordered so that items are always after any of their dependencies.
-func getDeps(m moduleName) []moduleName {
+// orderedDeps gets a list of all dependencies ordered so that items are always after any of their dependencies.
+func orderedDeps(m moduleName) []moduleName {
 	deps := listDeps(m)
 
 	// get a unique list of moduleNames, with a flag for whether they have been added to our result
-	uniq := make(map[moduleName]bool)
-	for _, d := range deps {
-		uniq[d] = false
+	uniq := map[moduleName]bool{}
+	for _, dep := range deps {
+		uniq[dep] = false
 	}
 
-	result := make([]moduleName, len(uniq))
-	numAdded := 0
+	result := make([]moduleName, 0)
 
 	// keep looping through all modules until they have all been added to the result.
-	for numAdded < len(uniq) {
-		for d, s := range uniq {
-			if s {
+
+	for len(result) < len(uniq) {
+	OUTER:
+		for name, added := range uniq {
+			if added {
 				continue
 			}
-			// check if all dependencies have been added to the result
-			allDepsAdded := true
-			for _, dep := range modules[d].deps {
+			for _, dep := range modules[name].deps {
+				// stop processing this module if one of its dependencies has
+				// not been added to the result yet.
 				if !uniq[dep] {
-					allDepsAdded = false
-					break
+					continue OUTER
 				}
 			}
 
-			// if all of the module's dependencies have been added the result slice,
+			// if all of the module's dependencies have been added to the result slice,
 			// then we can safely add this module to the result slice as well.
-			if allDepsAdded {
-				uniq[d] = true
-				result[numAdded] = d
-				numAdded++
-			}
+			uniq[name] = true
+			result = append(result, name)
 		}
 	}
 	return result

--- a/pkg/loki/modules.go
+++ b/pkg/loki/modules.go
@@ -54,7 +54,7 @@ func (m moduleName) String() string {
 	case All:
 		return "all"
 	default:
-		panic(fmt.Sprintf("unknow module name: %d", m))
+		panic(fmt.Sprintf("unknown module name: %d", m))
 	}
 }
 
@@ -176,6 +176,55 @@ func (t *Loki) initStore() (err error) {
 func (t *Loki) stopStore() error {
 	t.store.Stop()
 	return nil
+}
+
+// listDeps recursively gets a list of dependencies for a passed moduleName
+func listDeps(m moduleName) []moduleName {
+	deps := modules[m].deps
+	for _, d := range modules[m].deps {
+		deps = append(deps, listDeps(d)...)
+	}
+	return deps
+}
+
+// getDeps gets a list of all dependencies ordered so that items are always after any of their dependencies.
+func getDeps(m moduleName) []moduleName {
+	deps := listDeps(m)
+
+	// get a unique list of moduleNames, with a flag for whether they have been added to our result
+	uniq := make(map[moduleName]bool)
+	for _, d := range deps {
+		uniq[d] = false
+	}
+
+	result := make([]moduleName, len(uniq))
+	numAdded := 0
+
+	// keep looping through all modules until they have all been added to the result.
+	for numAdded < len(uniq) {
+		for d, s := range uniq {
+			if s {
+				continue
+			}
+			// check if all dependencies have been added to the result
+			allDepsAdded := true
+			for _, dep := range modules[d].deps {
+				if !uniq[dep] {
+					allDepsAdded = false
+					break
+				}
+			}
+
+			// if all of the module's dependencies have been added the result slice,
+			// then we can safely add this module to the result slice as well.
+			if allDepsAdded {
+				uniq[d] = true
+				result[numAdded] = d
+				numAdded++
+			}
+		}
+	}
+	return result
 }
 
 type module struct {

--- a/pkg/loki/modules.go
+++ b/pkg/loki/modules.go
@@ -197,7 +197,7 @@ func orderedDeps(m moduleName) []moduleName {
 		uniq[dep] = false
 	}
 
-	result := make([]moduleName, 0)
+	result := make([]moduleName, 0, len(uniq))
 
 	// keep looping through all modules until they have all been added to the result.
 

--- a/pkg/loki/modules_test.go
+++ b/pkg/loki/modules_test.go
@@ -6,7 +6,7 @@ import (
 
 func TestGetDeps(t *testing.T) {
 	for _, m := range []moduleName{All, Distributor, Ingester, Querier} {
-		deps := getDeps(m)
+		deps := orderedDeps(m)
 		seen := make(map[moduleName]struct{})
 		// make sure that getDeps always orders dependencies correctly.
 		for _, d := range deps {

--- a/pkg/loki/modules_test.go
+++ b/pkg/loki/modules_test.go
@@ -1,0 +1,21 @@
+package loki
+
+import (
+	"testing"
+)
+
+func TestGetDeps(t *testing.T) {
+	for _, m := range []moduleName{All, Distributor, Ingester, Querier} {
+		deps := getDeps(m)
+		seen := make(map[moduleName]struct{})
+		// make sure that getDeps always orders dependencies correctly.
+		for _, d := range deps {
+			seen[d] = struct{}{}
+			for _, dep := range modules[d].deps {
+				if _, ok := seen[dep]; !ok {
+					t.Errorf("module %s has dependency %s which has not been seen.", d, dep)
+				}
+			}
+		}
+	}
+}

